### DISCUSSION TITLE
rs232 to rs485 conversion

### DISCUSSION THREAD

**Discussion by travelo94** [micropython/micropython#10254] [RP2040 / Pico]
hi im using uart on pico with rs232 converter now i want to use rs485 , 
i wrote some demo code on rs232 that works.
`from machine import UART, Pin
 import machine
 uart=(0, baudrate=9600, bits=8, parity=None, stop=1, tx=Pin(0), rx=Pin(1), txbuf=256, rxbuf=256, timeout=2000)
 data=[0x00,0x01]
 uart.write(bytes(data))
 uart.read()

`
do i have to change anything to use rs485?  beacuse when i connected circuit it didnt worked

**Comment by karfas** [comment]
>  beacuse when i connected circuit it didnt worked

... is hardly a problem description someone can work with. 

You know that RS232/RS458/UART are completely different things electrically ?

**Reply by travelo94** [reply]
yes thats why the question is about the code 
do i have to change something in code to use rs485 to ttl converter or it can be the same as i was using rs232 to ttl

**Reply by rolandvs** [reply]
The rs485 is half-duplex. Provided your interface has no auto-select, you need to change/enable the driver. Signal nRE and DI are connected together so taking the pin low will enable the receiver and disable the transmitter and visa versa. Lookmin a datasheet of the MAX485 et all for more info.

ESP32/STM32 have dedicated pins to drive the line, but it depends on the software if they are supporting this. Anyway, you can toggle the line yourself. One caution, make sure the transmit buffer is empty before switching.

R

**Reply by travelo94** [reply]
okey so i change the pin that is connected to rse[de+re] on converter on high before sending data and on low before reading?
`from machine import UART, Pin
import machine
uart=(0, baudrate=9600, bits=8, parity=None, stop=1, tx=Pin(0), rx=Pin(1), txbuf=256, rxbuf=256, timeout=2000)
rse=Pin(15, Pin.OUT)
data=[0x00,0x01]
rse.high()
uart.write(bytes(data))
rse.low()
uart.read()`
connecting gnd of the device with rs485 to the pico gnd  and a , b to a , b of the converter is correct tho?

**Reply by robert-hh** [reply]
Which converter do you use?

> One caution, make sure the transmit buffer is empty before switching.

The UART output of the ESP32 is buffered. The call to uart.write() will for messages shorter than the buffer size return immediately before the data has been sent. The default buffer size is 256. You can call uart.flush() to wait, until all data has been sent, or use uart.txdone() to poll, whether all data has been sent, at which time uart.txdone() will return True. I assume you intended to write:
```
from machine import UART, Pin
import machine
uart=UART(0, baudrate=9600, bits=8, parity=None, stop=1, tx=Pin(0), rx=Pin(1), txbuf=256, rxbuf=256, timeout=2000) 
rse=Pin(15, Pin.OUT)
data=[0x00,0x01]
rse.high()
uart.write(bytes(data))
uart.flush()  # Added to wait for the data being sent
rse.low()
uart.read()
```

**Reply by travelo94** [reply]
i use waveshare 4738 
I added the uart.flush() i got error AttributeError: UART object has no attribute 'flush'
same error with txdone()

**Reply by robert-hh** [reply]
Please update the firmware to a recent daily build. uart.flush() was added after v1.19.1.

**Reply by robert-hh** [reply]
Script updated, not using UART 0 and pins 0 and 1, which does not work here
```
from machine import UART, Pin
import machine
uart=UART(1, baudrate=9600, bits=8, parity=None, stop=1, tx=4, rx=5, txbuf=256, rxbuf=256, timeout=2000) 
rse=Pin(15, Pin.OUT)
data=[0x00,0x01]
rse.high()
uart.write(bytes(data))
uart.flush()  # Added to wait for the data being sent
rse.low()
uart.read()
```

**Reply by travelo94** [reply]
i use thonny 4.0.1 there is no newer version

**Reply by robert-hh** [reply]
I talked about the firmware of the ESP32 board, and you do not need thonny top update it. The fESP32 firmware is at https://micropython.org/download/?port=esp32 together with update instructions. if you use Windows or MAC OSx, you have to use a different USB port name.

**Comment by robert-hh** [comment]
Glancing into Thonny I see, that you can use it to update the ESP32 firmware, after having downloaded the proper version.

**Reply by travelo94** [reply]
> uart.flush() was added after v1.19.1. >
including v1.19.1? Beacuse it is  the newest version here:
https://micropython.org/download/rp2-pico/
I updated to this version : MicroPython v1.19.1 on 2022-06-18; Raspberry Pi Pico withRP2040
and still got that error

**Reply by robert-hh** [reply]
Ok. You're using RPi. Somehow I thought it was ESP32. Nevertheless, use version: [v1.19.1-776-gbb77c1d5a (2022-12-16) .uf2](https://micropython.org/resources/firmware/rp2-pico-20221216-unstable-v1.19.1-776-gbb77c1d5a.uf2) from the nightly builds.
And with RPi you can indeed use UART 0 at pins 0 and 1.
A note about uart.flush(): You have to add a 1 character delay after uart.flush() returns, since it returns while the last byte is sent, not after the last byte was sent. At 9600 baud, a 1 character delay is about 1 ms.

**Reply by travelo94** [reply]
used v1.19.1-776 and dont threw an error changed the uart to 1 and still dont work
i connected 
converter A to device A 
converter B to device B
device GND to 5v power supply GND
Rse to pin 15
DI to tx 
RO to rx
VCC and GND of converter to 5v power supply 
and pico gnd right after tx,rx to 5v power supply gnd
`from machine import UART, Pin
import machine
uart=UART(1, baudrate=9600, bits=8, parity=None, stop=1, tx=4, rx=5, txbuf=256, rxbuf=256, timeout=2000) 
rse=Pin(15, Pin.OUT)
data=[0x00,0x01]
rse.high()
uart.write(bytes(data))
uart.flush()  # Added to wait for the data being sent
time.sleep_ms(1)
rse.low()
print(uart.read())`

**Reply by robert-hh** [reply]
```
from machine import UART, Pin
import machine
import time
uart=UART(0, baudrate=9600, bits=8, parity=None, stop=1, tx=Pin(0), rx=Pin(1), txbuf=256, rxbuf=256, timeout=2000) 
rse=Pin(15, Pin.OUT)
data=[0x00,0x01]
rse.high()
uart.write(bytes(data))
uart.flush()  # Added to wait for the data being sent
time.sleep_ms(1)
rse.low()
result = uart.read()
print(result)
 ```

**Reply by travelo94** [reply]
yes i added the delay in the comment higher and still something is off thats why i added how i connected the circuit

**Reply by robert-hh** [reply]
Seems similar. Did you connect rse to GPIO15, which is board pin 20, or to board pin 15?

**Reply by robert-hh** [reply]
The delay must be just right, otherwise the received data is lost. You can try small changes using sleep_us().

**Reply by robert-hh** [reply]
GPIO4 is board pin 6, GPIO5 is board pin 7. Micropython uses GPIO numbers, not board pin numbers,

**Reply by travelo94** [reply]
ye the gpio15 the last one on the left side i was even checking and added led to see if the state is changing 
checked all delays from 100us to 3000 and still

**Reply by karfas** [reply]
I would check on hardware level where the signal gets lost. 
Just add a LED (maybe with an resistor) 
- from UART tx to ground 
- between input of the RS485 to ground
- between the RS485 lines

You will see the led flickering when you send data and/or see it flickering in a different way when one of the attached devices returns something on the RS485 bus.
